compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp

```diff
@@ -1608,23 +1609,55 @@ LogicalResult WinogradOutputTransformOp::getResultTilePosition(
 //===----------------------------------------------------------------------===//
 
 SmallVector<Range> AttentionOp::getIterationDomain(OpBuilder &builder) {
-  int64_t iterationDomainRank = getIterationDomainRank();
-  SmallVector<Range> loopBounds(iterationDomainRank);
+  int64_t domainRank = getIterationDomainRank();
+
+  SmallVector<Range> loopBounds(domainRank);
   Location loc = getLoc();
   Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
   Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
-  Value source = getQuery();
-  for (auto dim : llvm::seq<int64_t>(0, iterationDomainRank)) {
+
+  for (auto dim : llvm::seq<int64_t>(0, domainRank)) {
     loopBounds[dim].offset = zero;
-    loopBounds[dim].size = getDimValue(builder, loc, source, dim);
     loopBounds[dim].stride = one;
   }
+
+  SmallVector<bool> dimsFound(domainRank, false);
+  auto fillSizes = [&](Value val, AffineMap indexingMap) {
+    for (auto [idx, dimExpr] : llvm::enumerate(indexingMap.getResults())) {
+      assert(isa<AffineDimExpr>(dimExpr));
+      AffineDimExpr dim = cast<AffineDimExpr>(dimExpr);
+      int64_t pos = dim.getPosition();
+      if (dimsFound[pos]) {
+        continue;
+      }
+      dimsFound[pos] = true;
+      loopBounds[pos].size = getDimValue(builder, loc, val, idx);
+    }
+  };
+
+  // Sizes can be found from Q, K, V alone.
+  fillSizes(getQuery(), getQueryMap());
+  fillSizes(getKey(), getKeyMap());
+  fillSizes(getValue(), getValueMap());
+
   return loopBounds;
 }
 
 SmallVector<utils::IteratorType> AttentionOp::getLoopIteratorTypes() {
+  FailureOr<AttentionOpDetail> maybeOpInfo =
+      AttentionOpDetail::get(getIndexingMapsArray());
+  assert(succeeded(maybeOpInfo) && "Failed to infer attention op details");
+  AttentionOpDetail opInfo = maybeOpInfo.value();
+
+  // All dimensions other than k1 and k2 are parallel.
   SmallVector<utils::IteratorType> iteratorTypes(getIterationDomainRank(),
                                                  utils::IteratorType::parallel);
+
+  for (int dim :
+       llvm::concat<const int64_t>(opInfo.getK1Dims(), opInfo.getK2Dims())) {
+    iteratorTypes[dim] = utils::IteratorType::reduction;
+  }
```

Comment:
The type mismatch, perhaps we just use `auto` for `dim`. Because `concat<>` already spells the type.

Reply:
done